Content.Shared/TapeRecorder/SharedTapeRecorderSystem.cs

```diff
@@ -0,0 +1,610 @@
+
+        return true;
+    }
+
+    /// <summary>
+    /// Stop rewinding if we are rewinding
+    /// </summary>
+    protected virtual bool Stop(Entity<TapeRecorderComponent> tapeRecorder, EntityUid? user = null, bool updateAppearance = true, bool changeMode = false)
+    {
+        if (!tapeRecorder.Comp.Active)
+            return false;
+
+        if (changeMode)
+            SetMode(tapeRecorder, TapeRecorderMode.Stopped);
+
+        SetActive(tapeRecorder, false);
+
+        if (updateAppearance)
+            UpdateAppearance(tapeRecorder, TapeRecorderMode.Stopped);
+
+        tapeRecorder.Comp.NeedUIUpdate = true;
+
+        _audioSystem.PlayPvs(tapeRecorder.Comp.StopSound, tapeRecorder);
+
+        return true;
+    }
+
+    /// <summary>
+    /// Set the tape recorder mode and dirty if it is different from the previous mode
+    /// </summary>
+    /// <param name="tapeRecorder">The tape recorder to update</param>
+    /// <param name="mode">The new mode</param>
+    /// <param name="updateAppearance">Should the appearance of the tape recorder be updated</param>
+    protected void SetMode(Entity<TapeRecorderComponent> tapeRecorder, TapeRecorderMode mode)
+    {
+        if (mode == tapeRecorder.Comp.Mode)
+            return;
+
+        tapeRecorder.Comp.Mode = mode;
+
+        Dirty(tapeRecorder);
+    }
+
+    protected void SetActive(Entity<TapeRecorderComponent> tapeRecorder, bool active)
+    {
+        if (active == tapeRecorder.Comp.Active)
+            return;
+
+        tapeRecorder.Comp.Active = active;
+
+        Dirty(tapeRecorder);
+    }
+
+    protected bool TryGetTapeCassette(EntityUid tapeRecorder, [NotNullWhen(true)] out Entity<TapeCassetteComponent> tapeCassette)
+    {
+        var cassette = _itemSlotsSystem.GetItemOrNull(tapeRecorder, SlotName);
+        if (cassette == null)
+        {
+            tapeCassette = default!;
+            return false;
+        }
+
+        if (!TryComp<TapeCassetteComponent>(cassette, out var tapeCassetteComponent))
+        {
+            tapeCassette = default!;
+            return false;
+        }
+
+        tapeCassette = new(cassette.Value, tapeCassetteComponent);
+
+        return true;
+    }
+
+    private void UpdateUI(Entity<TapeRecorderComponent> tapeRecorder)
+    {
+        if (!_uiSystem.IsUiOpen(tapeRecorder.Owner, TapeRecorderUIKey.Key))
+            return;
+
+        var comp = tapeRecorder.Comp;
+        if (comp == null)
+            return;
+
+        var hasCassette = TryGetTapeCassette(tapeRecorder, out var cassette);
+        var hasData = false;
+        var currentTime = 0f;
+        var maxTime = 0f;
+        var cassetteName = "Unnamed";
+        var cooldown = comp.PrintCooldown;
+
+        if (hasCassette)
+        {
+            hasData = cassette.Comp.RecordedData.Any();
+            currentTime = cassette.Comp.CurrentPosition;
+            maxTime = (float)cassette.Comp.MaxCapacity.TotalSeconds;
```

Comment:
This should be `TimeSpan` instead of `float`, you shouldn't need a cast.

Reply:
Can't understand why i need timespan here, i just need it for Slider's maxvalue